p2p/src/net/libp2p/mod.rs

```diff
@@ -179,21 +197,65 @@ where
         .collect::<Vec<net::AddrInfo<T>>>()
 }
 
+impl<T> TryInto<net::PeerInfo<T>> for IdentifyInfo
+where
+    T: NetworkService<PeerId = PeerId, ProtocolId = String>,
+{
+    type Error = P2pError;
+
+    fn try_into(self) -> Result<net::PeerInfo<T>, Self::Error> {
+        let proto = self.protocol_version.clone();
+        let (version, net) =
+            match sscanf::scanf!(proto, "/{}/{}.{}.{}-{}", String, u8, u8, u16, String) {
+                Err(e) => Err(P2pError::ProtocolError(ProtocolError::InvalidProtocol)),
+                Ok((proto, maj, min, pat, magic)) => {
+                    if proto != "mintlayer" {
+                        return Err(P2pError::ProtocolError(ProtocolError::InvalidProtocol));
+                    }
+
+                    // TODO: `impl TryInto<ChainType> for u32`?
+                    let net = match magic.as_str() {
+                        "1a64e5f1" => common::chain::config::ChainType::Mainnet,
+                        _ => return Err(P2pError::ProtocolError(ProtocolError::UnknownNetwork)),
+                    };
```

Comment:
Sure, I'll fix that. The problem I have the chainconfig is that there's a separate entry for the chaintype and magic but IIUC they should be coupled together so we either don't need to number because the enum value is unique or we have some conversion from the enum value to an `u32` if we truly need this magic number for something. Alternatively we don't add the magic number to the protocol string and instead let the `SwarmManager` decided which connections are kept open and which are closed. I'm personally leaning towards not having the magic number in the protocol string and instead using `/mintlayer/0.1.0` and `/mintlayer-testnet/0.1.0` to distinguish mainnet and testnet.

The point I'm trying to make is that in its current form, the code is pretty ugly because we store the magic number to the protocol string and then extract that string and try to convert it to a `ChainType`. Not entirely sure how to fix it so I'm open to suggestions.